# Project: Market Segmentation
**Notebook 2: Modelling, Conclusions and Recommendation**

## TABLE OF CONTENTS

[(1) Modelling](#1.-Modelling) <br>
[(2) Conclusions](#2.-Conclusions) <br>
[(3) Recommendation](#3.-Recommendation) <br>

# Modelling

In this section, the following process will be carried out:

- [Set up Data for Modelling](#Set-up-Data-for-Modelling) <br>
- [Split Data into Training and Testing Sets](#Split-Data-into-Training-and-Testing-Sets) <br>
- [Training and Testing Model Accuracy using Decision Tree](#Training-and-Testing-Model-Accuracy-using-Decision-Tree) <br>

## Set up Data for Modelling

### Import Libraries

In [1]:
# Import Libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix
from sklearn import metrics
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

In [2]:
# Show All Rows in Dataset
pd.set_option('display.max_rows', None)

# Show All Column in Dataset
pd.set_option('display.max_columns', None)

### Import Data

In [3]:
# Read data from csv.
train_dataframe = pd.read_csv("./datasets/train_dataframe_cluster.csv")

### Data Overview

In [4]:
# Summary of dataFrame.
print("***************************************")
print("     Summary of Insurance Dataframe    ")
print("***************************************")
train_dataframe.info()

***************************************
     Summary of Insurance Dataframe    
***************************************
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59381 entries, 0 to 59380
Columns: 146 entries, Unnamed: 0 to Product_Info_2_E1
dtypes: float64(18), int64(128)
memory usage: 66.1 MB


**Observation**:
- There are 128 columns/features and 59,380 rows in this dataframe.
- There are 127 columns/features with numerical data (109 columns/features with int64 data type; and 18 columns/features with float64 data type) and 1 column/feature with categorical data.

In [5]:
# Print shape of dataframe.
print(f"Shape:", train_dataframe.shape)

Shape: (59381, 146)


In [6]:
# Displays first 5 rows of dataframe.
train_dataframe.head()

Unnamed: 0  Product_Info_1  Product_Info_3  Product_Info_4  Product_Info_5  \
0           0               1              10        0.076923               2   
1           1               1              26        0.076923               2   
2           2               1              26        0.076923               2   
3           3               1              10        0.487179               2   
4           4               1              26        0.230769               2   

   Product_Info_6  Product_Info_7   Ins_Age        Ht        Wt       BMI  \
0               1               1  0.641791  0.581818  0.148536  0.323008   
1               3               1  0.059701  0.600000  0.131799  0.272288   
2               3               1  0.029851  0.745455  0.288703  0.428780   
3               3               1  0.164179  0.672727  0.205021  0.352438   
4               3               1  0.417910  0.654545  0.234310  0.424046   

   Employment_Info_1  Employment_Info_2  Employment_Info_3  Employment_Info_4  \
0              0.028                 12                  1                0.0   
1              0.000                  1                  3                0.0   
2              0.030                  9                  1                0.0   
3              0.042                  9                  1                0.0   
4              0.027                  9                  1                0.0   

   Employment_Info_5  Employment_Info_6  InsuredInfo_1  InsuredInfo_2  \
0                  3           0.361469              1              2   
1                  2           0.001800              1              2   
2                  2           0.030000              1              2   
3                  3           0.200000              2              2   
4                  2           0.050000              1              2   

   InsuredInfo_3  InsuredInfo_4  InsuredInfo_5  InsuredInfo_6  InsuredInfo_7  \
0              6              3              1              2              1   
1              6              3              1              2              1   
2              8              3              1              1              1   
3              8              3              1              2              1   
4              6              3              1              2              1   

   Insurance_History_1  Insurance_History_2  Insurance_History_3  \
0                    1                    1                    3   
1                    2                    1                    3   
2                    2                    1                    1   
3                    2                    1                    1   
4                    2                    1                    1   

   Insurance_History_4  Insurance_History_5  Insurance_History_7  \
0                    1             0.000667                    1   
1                    1             0.000133                    1   
2                    3             0.001733                    3   
3                    3             0.001733                    3   
4                    3             0.001733                    3   

   Insurance_History_8  Insurance_History_9  Family_Hist_1  Family_Hist_2  \
0                    1                    2              2       0.474550   
1                    3                    2              2       0.188406   
2                    2                    3              3       0.304348   
3                    2                    3              3       0.420290   
4                    2                    3              2       0.463768   

   Family_Hist_3  Family_Hist_4  Family_Hist_5  Medical_History_1  \
0       0.598039       0.444890       0.526786           4.000000   
1       0.497737       0.084507       0.484635           5.000000   
2       0.497737       0.225352       0.484635          10.000000   
3       0.497737       0.352113       0.484635           0.000000   
4       0.497737    

In [7]:
# Displays last 5 rows of dataframe.
train_dataframe.tail()

Unnamed: 0  Product_Info_1  Product_Info_3  Product_Info_4  \
59376       59376               1              10        0.230769   
59377       59377               1              26        0.230769   
59378       59378               1              26        0.076923   
59379       59379               1              10        0.230769   
59380       59380               1              26        0.076923   

       Product_Info_5  Product_Info_6  Product_Info_7   Ins_Age        Ht  \
59376               2               3               1  0.074627  0.709091   
59377               2               3               1  0.432836  0.800000   
59378               2               3               1  0.104478  0.745455   
59379               2               3               1  0.507463  0.690909   
59380               2               3               1  0.447761  0.781818   

             Wt       BMI  Employment_Info_1  Employment_Info_2  \
59376  0.320084  0.519103              0.020                  1   
59377  0.403766  0.551119              0.100                  9   
59378  0.246862  0.360969              0.035                  9   
59379  0.276151  0.462452              0.038                  9   
59380  0.382845  0.539563              0.123                  9   

       Employment_Info_3  Employment_Info_4  Employment_Info_5  \
59376                  3           0.000000                  3   
59377                  1           0.000010                  2   
59378                  1           0.000000                  2   
59379                  1           0.006283                  3   
59380                  1           0.006283                  2   

       Employment_Info_6  InsuredInfo_1  InsuredInfo_2  InsuredInfo_3  \
59376           0.025000              1              2              8   
59377           0.350000              1              2              3   
59378           0.361469              1              2              6   
59379           0.361469              1              2              3   
59380           0.300000              1              2              6   

       InsuredInfo_4  InsuredInfo_5  InsuredInfo_6  InsuredInfo_7  \
59376              3              1              2              1   
59377              3              1              1              1   
59378              3              1              1              1   
59379              3              1              2              1   
59380              3              1              1              1   

       Insurance_History_1  Insurance_History_2  Insurance_History_3  \
59376                    2                    1                    1   
59377                    2                    1                    3   
59378                    2                    1                    1   
59379                    2                    1                    1   
59380                    2                    1                    1   

       Insurance_History_4  Insurance_History_5  Insurance_History_7  \
59376                    3             0.001733                    3   
59377                    2             0.000267                    1   
59378                    3             0.001733                    3   
59379                    3             0.001733                    3   
59380                    3             0.001733                    3   

       Insurance_History_8  Insurance_History_9  Family_Hist_1  Family_Hist_2  \
59376                    2                    3              3       0.217391   
59377                    3                    2              3       0.565217   
59378                    2                    3              3       0.173913   
59379                    2                    3              2       0.474550   
59380                    2                    3              2       0.474550   

       Family_Hist_3  Family_Hist_4  Family_Hist_5  Medical_History_1  \
59376       0.497737       0.197183       0.48463

### Select Features

All columns from train_dataframe will be selected as features except `cluster`.

In [8]:
# Drop both "Unnamed: 0" and "cluster" and select the remaining columns as features.
selected_features = train_dataframe.drop(['Unnamed: 0', 'cluster'], axis=1)
selected_features.head()

Product_Info_1  Product_Info_3  Product_Info_4  Product_Info_5  \
0               1              10        0.076923               2   
1               1              26        0.076923               2   
2               1              26        0.076923               2   
3               1              10        0.487179               2   
4               1              26        0.230769               2   

   Product_Info_6  Product_Info_7   Ins_Age        Ht        Wt       BMI  \
0               1               1  0.641791  0.581818  0.148536  0.323008   
1               3               1  0.059701  0.600000  0.131799  0.272288   
2               3               1  0.029851  0.745455  0.288703  0.428780   
3               3               1  0.164179  0.672727  0.205021  0.352438   
4               3               1  0.417910  0.654545  0.234310  0.424046   

   Employment_Info_1  Employment_Info_2  Employment_Info_3  Employment_Info_4  \
0              0.028                 12                  1                0.0   
1              0.000                  1                  3                0.0   
2              0.030                  9                  1                0.0   
3              0.042                  9                  1                0.0   
4              0.027                  9                  1                0.0   

   Employment_Info_5  Employment_Info_6  InsuredInfo_1  InsuredInfo_2  \
0                  3           0.361469              1              2   
1                  2           0.001800              1              2   
2                  2           0.030000              1              2   
3                  3           0.200000              2              2   
4                  2           0.050000              1              2   

   InsuredInfo_3  InsuredInfo_4  InsuredInfo_5  InsuredInfo_6  InsuredInfo_7  \
0              6              3              1              2              1   
1              6              3              1              2              1   
2              8              3              1              1              1   
3              8              3              1              2              1   
4              6              3              1              2              1   

   Insurance_History_1  Insurance_History_2  Insurance_History_3  \
0                    1                    1                    3   
1                    2                    1                    3   
2                    2                    1                    1   
3                    2                    1                    1   
4                    2                    1                    1   

   Insurance_History_4  Insurance_History_5  Insurance_History_7  \
0                    1             0.000667                    1   
1                    1             0.000133                    1   
2                    3             0.001733                    3   
3                    3             0.001733                    3   
4                    3             0.001733                    3   

   Insurance_History_8  Insurance_History_9  Family_Hist_1  Family_Hist_2  \
0                    1                    2              2       0.474550   
1                    3                    2              2       0.188406   
2                    2                    3              3       0.304348   
3                    2                    3              3       0.420290   
4                    2                    3              2       0.463768   

   Family_Hist_3  Family_Hist_4  Family_Hist_5  Medical_History_1  \
0       0.598039       0.444890       0.526786           4.000000   
1       0.497737       0.084507       0.484635           5.000000   
2       0.497737       0.225352       0.484635          10.000000   
3       0.497737       0.352113       0.484635           0.000000   
4       0.497737       0.408451       0.484635           7.962172   

   Medical_History_2  

In [9]:
# `X` will be the `selected_features` columns.
# `y` will be the `cluster` column.
X = selected_features
y = train_dataframe['cluster']

## Split Data into Training and Testing Sets

In [10]:
# Split the data into the training and testing sets.
X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size=0.30,   # 30% of data is retained for testing.
                                                    random_state=42)

In [11]:
# Display shape of X_train and X_test.
# Print X_train shape.
print(f"X_train shape:", X_train.shape)
# Print X_test shape.
print(f"X_test shape:", X_test.shape)

X_train shape: (41566, 144)
X_test shape: (17815, 144)


In [12]:
# Display first 5 rows of X_train.
X_train.head()

Product_Info_1  Product_Info_3  Product_Info_4  Product_Info_5  \
4010                1              10        0.076923               2   
48030               1              26        0.102564               2   
38496               1              26        0.230769               2   
12290               1              26        0.282051               2   
35280               1              26        0.128205               2   

       Product_Info_6  Product_Info_7   Ins_Age        Ht        Wt       BMI  \
4010                1               1  0.671642  0.618182  0.196653  0.387759   
48030               3               1  0.701493  0.745455  0.261506  0.384703   
38496               3               1  0.119403  0.618182  0.198745  0.391944   
12290               1               1  0.567164  0.690909  0.215481  0.355123   
35280               1               1  0.761194  0.763636  0.414226  0.609455   

       Employment_Info_1  Employment_Info_2  Employment_Info_3  \
4010               0.000                  1                  3   
48030              0.010                 14                  1   
38496              0.046                  9                  1   
12290              0.130                  9                  1   
35280              0.100                  9                  1   

       Employment_Info_4  Employment_Info_5  Employment_Info_6  InsuredInfo_1  \
4010            0.013548                  3              0.010              2   
48030           0.120000                  2              1.000              1   
38496           0.000000                  2              0.045              1   
12290           0.007200                  2              0.500              1   
35280           0.000000                  2              0.500              1   

       InsuredInfo_2  InsuredInfo_3  InsuredInfo_4  InsuredInfo_5  \
4010               2              8              3              1   
48030              2              3              3              1   
38496              2              3              3              1   
12290              2              3              3              1   
35280              2              8              3              1   

       InsuredInfo_6  InsuredInfo_7  Insurance_History_1  Insurance_History_2  \
4010               2              1                    1                    1   
48030              1              1                    2                    1   
38496              2              1                    2                    1   
12290              1              1                    2                    1   
35280              1              1                    1                    1   

       Insurance_History_3  Insurance_History_4  Insurance_History_5  \
4010                     3                    1             0.000667   
48030                    3                    1             0.000900   
38496                    3                    2             0.000313   
12290                    3                    1             0.002000   
35280                    3                    1             0.000920   

       Insurance_History_7  Insurance_History_8  Insurance_History_9  \
4010                     2                    1                    2   
48030                    1                    3                    1   
38496                    1                    3                    2   
12290                    1                    3                    2   
35280                    2                    1                    2   

       Family_Hist_1  Family_Hist_2  Family_Hist_3  Family_Hist_4  \
4010               3       0.474550       0.735294       0.444890   
48030              3       0.474550       0.676471       0.444890   
38496              3       0.246377       0.497737       0.444890   
12290              3       0.710145       0.497737       0.605634   
35280              2       0.474550       0.490196       0.444890   

       Family_

In [13]:
# Display first 5 rows of X_test.
X_test.head()

Product_Info_1  Product_Info_3  Product_Info_4  Product_Info_5  \
41952               1              26        0.076923               2   
28556               1              26        0.076923               2   
7984                1              26        0.487179               2   
48954               1              10        0.230769               2   
40991               1              26        0.230769               2   

       Product_Info_6  Product_Info_7   Ins_Age        Ht        Wt       BMI  \
41952               3               1  0.761194  0.745455  0.292887  0.435562   
28556               1               1  0.313433  0.636364  0.177824  0.333602   
7984                3               1  0.208955  0.745455  0.560669  0.869557   
48954               3               1  0.462687  0.709091  0.177824  0.274782   
40991               3               1  0.223881  0.654545  0.330544  0.604930   

       Employment_Info_1  Employment_Info_2  Employment_Info_3  \
41952              0.045                 14                  1   
28556              0.040                 12                  1   
7984               0.060                  9                  1   
48954              0.035                  9                  1   
40991              0.024                  9                  1   

       Employment_Info_4  Employment_Info_5  Employment_Info_6  InsuredInfo_1  \
41952           0.006283                  2           0.100000              2   
28556           0.000000                  2           0.015000              2   
7984            0.000000                  2           0.361469              1   
48954           0.000000                  3           0.361469              1   
40991           0.000000                  2           0.361469              2   

       InsuredInfo_2  InsuredInfo_3  InsuredInfo_4  InsuredInfo_5  \
41952              2             11              2              1   
28556              2              8              3              1   
7984               2              6              3              1   
48954              2              3              3              1   
40991              2              2              3              1   

       InsuredInfo_6  InsuredInfo_7  Insurance_History_1  Insurance_History_2  \
41952              1              1                    2                    1   
28556              2              1                    2                    1   
7984               1              1                    2                    1   
48954              2              1                    2                    1   
40991              2              1                    2                    1   

       Insurance_History_3  Insurance_History_4  Insurance_History_5  \
41952                    1                    3             0.001733   
28556                    1                    3             0.001733   
7984                     1                    3             0.001733   
48954                    1                    3             0.001733   
40991                    1                    3             0.001733   

       Insurance_History_7  Insurance_History_8  Insurance_History_9  \
41952                    3                    2                    3   
28556                    3                    2                    3   
7984                     3                    2                    3   
48954                    3                    2                    3   
40991                    3                    2                    3   

       Family_Hist_1  Family_Hist_2  Family_Hist_3  Family_Hist_4  \
41952              3       0.474550       0.497737       0.444890   
28556              3       0.579710       0.497737       0.464789   
7984               2       0.492754       0.497737       0.444890   
48954              3       0.521739       0.497737       0.492958   
40991              3       0.474550       0.352941       0.267606   

       Family_

## Training and Testing Model Accuracy using Decision Tree

In [14]:
model= DecisionTreeClassifier(criterion="entropy")
model.fit(X_train, y_train)

DecisionTreeClassifier(criterion='entropy')

In [15]:
# Evaluate accuracy on training data
train_acc = model.score(X_train, y_train)
print("Training accuracy:", train_acc)

Training accuracy: 1.0


In [16]:
# Evaluate accuracy on testing data
test_acc = model.score(X_test, y_test)
print("Testing accuracy:", test_acc)

Testing accuracy: 1.0


In [17]:
y_pred = model.predict(X_test)

In [18]:
#Confusion_Matrix
print(metrics.confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[11490     0]
 [    0  6325]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     11490
           1       1.00      1.00      1.00      6325

    accuracy                           1.00     17815
   macro avg       1.00      1.00      1.00     17815
weighted avg       1.00      1.00      1.00     17815



**Observation**:
- This model is accurate with accuracy score at 100%.

# Conclusions

Customers are segregated into 2 clusters. Among all, cluster 0 has the highest count of data with cluster 1 having the lowest count of data.

# Recommendation

As the data has been normalized and dummified, it is challenging to derive meaningful insights from the data for product and campaign development for the targeted clusters. Hence, for future improvement, clarification could be seek from Prudential to understand the data and then to derive actionable insights; and to continue training the model with more data points.